account_invoice_section_sale_order/models/sale_order.py

```diff
@@ -16,42 +18,53 @@ def _create_invoices(self, grouped=False, final=False):
         """
         invoice_ids = super()._create_invoices(grouped=grouped, final=final)
         for invoice in invoice_ids:
-            if len(invoice.line_ids.mapped("sale_line_ids.order_id.id")) == 1:
+            if (
+                len(invoice.line_ids.mapped(invoice.line_ids._get_section_grouping()))
+                == 1
+            ):
                 continue
-            so = None
             sequence = 10
+            move_lines = invoice._get_ordered_invoice_lines()
+            # Group move lines according to their sale order
+            section_grouping_matrix = OrderedDict()
+            for move_line in move_lines:
+                group = move_line._get_section_group()
+                group_move_line_ids = section_grouping_matrix.get(group)
+                if not group_move_line_ids:
+                    section_grouping_matrix[group] = [move_line.id]
+                else:
+                    group_move_line_ids.append(move_line.id)
+            # Prepare section lines for each group
             section_lines = []
-            lines = self._get_ordered_invoice_lines(invoice)
-            for line in lines:
-                if line.sale_line_ids.order_id and so != line.sale_line_ids.order_id:
-                    so = line.sale_line_ids.order_id
-                    section_lines.append(
-                        (
-                            0,
-                            0,
-                            {
-                                "name": so._get_saleorder_section_name(),
-                                "display_type": "line_section",
-                                "sequence": sequence,
-                            },
-                        )
+            for group, move_line_ids in section_grouping_matrix.items():
+                section_lines.append(
+                    (
+                        0,
+                        0,
+                        {
+                            "name": group._get_invoice_section_name(),
+                            "display_type": "line_section",
+                            "sequence": sequence,
+                        },
                     )
-                    sequence += 10
-                line.sequence = sequence
+                )
                 sequence += 10
+                for move_line in self.env["account.move.line"].browse(move_line_ids):
```

Comment:
interpolating can be slow on large recordsets. In this case is fine. Also, these records will be loaded from cache.

Reply:
Interpolating seems to be a kind of estimation at finding new data point are we sure this is related ?
I am surprised browsing for a record is not slower.
Also about performance for this, the implementation goes from one  loop to 3 indented loops.

Reply:
well, I assumed you wanted to collect move line line `mapping[group] |= line` but if you just append the to a list is fine.

Reply:
I'm not sure if the performance is really degraded through this refactor and would be interesting to know. As browsing should read from the cache since we already had to read these records, I chose to keep the ids because unions of recordset is usually more costly.

Then from my POV the code is still quite readable between:
1. Grouping of lines
2. Preparation of section lines to create
3. Update of the existing lines sequence
4. Creation of section lines

If performance is really an issue, I guess we can revisit it.